### PR TITLE
[Feat] Optional update of IG:dm

### PR DESCRIPTION
**Pull request by vineetvk01**
**Feature:** Enable/Disable check for auto-update.

**Description:** Auto-update to the new release is always _ON_ by default. This feature checks saved user-preference and updates only when Auto-update is enabled. 

**Preview:**

auto-update

__** WHATS NEW ? **__
**Storage API:** Introducing to storage.js which can help developer to store user preferences and can create new store to save/preserve data specific to each user.  

Closes: #1240 , #444

### CODE REVIEW

File: main/menutemplate.js [hunk 1]
```diff
@@ -108,16 +109,22 @@ if (process.platform === 'darwin') {
 
 const createMenuTemplate = () => {
   return new Promise((resolve) => {
-    autoLaunch.autoLaunchStatus().then((status) => {
+    autoLaunch.autoLaunchStatus().then((autoLaunchStatus) => {
       template.push(
         {
           label: 'Configurations',
```
Comment (ifedapoolarewaju) [inline main/menutemplate.js]: Oh I just realised that we  are `pushing` into the `template` array here. In Mac and windows, the "Help" menu is always the last element in the menu bar. Is there a way we can preserve that?

Comment (vineetvk01) [reply main/menutemplate.js]: Yes. Instead of pushing the "Configurations" Menu. I can splice it in between before the "Window" and "Help".

Comment (vineetvk01) [reply main/menutemplate.js]: Hey! @ifedapoolarewaju . I thought of inserting the "Configurations" before the last two menuItems using splice(-2, 0, { object } );

File: main/storage.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    return JSON.parse(fs.readFileSync(filePath));
+  } catch (error) {
+    // if there was some kind of error, return the passed in defaults instead.
+    return defaults;
+  }
+}
+
+const Store = function (storeProps) {
+  const userDataPath = (electron.app || electron.remote.app).getPath('userData');
+  this.path = path.join(userDataPath, storeProps.configName + '.json');
```
Comment (ifedapoolarewaju) [inline main/storage.js]: I'd rather we don't append `.json` to the names, so it doesn't by any chance clash with [user session file which we save in json files](https://github.com/igdmapps/igdm/blob/master/main/utils.js#L41)

File: main/storage.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+const electron = require('electron');
+const path = require('path');
+const fs = require('fs');
+
+function parseDataFile (filePath, defaults) {
+  // We'll try/catch it in case the file doesn't exist yet, which will be the case on the first application run.
+  // `fs.readFileSync` will return a JSON string which we then parse into a Javascript object
+  try {
+    return JSON.parse(fs.readFileSync(filePath));
```
Comment (ifedapoolarewaju) [inline main/storage.js]: before making calls to any `fs` operations, I think we should do a check similar [to this](https://github.com/igdmapps/igdm/blob/master/main/utils.js#L39). To see if the system will permit file persistence by the IGdm app. If it doesn't permit, then we will just use default settings. What do you think?

File: main/storage.js (hunk 2 above)
Comment (vineetvk01) [reply main/storage.js]: updated the extension as `.store`

File: main/storage.js (hunk 3 above)
Comment (vineetvk01) [reply main/storage.js]: Yes. Agree. Will check with utils.canUseFileStorage() before accessing the file.

File: main/storage.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  } catch (error) {
+    // if there was some kind of error, return the passed in defaults instead.
+    return defaults;
+  }
+}
+
+const Store = function (storeProps) {
```
[the remark this answers is not in the document]
Comment (vineetvk01) [reply main/storage.js]: As it was not used before in the project, I thought of not using class. But if its okay. i have applied the changes. :100:

**Comment by ifedapoolarewaju** [comment]
thank you!

**Pull request merged**